CRM/Price/BAO/LineItem.php

```diff
@@ -819,60 +825,107 @@ protected function _getFinancialItemsToRecord($entityID, $entityTable, $contribu
         // reverse the related financial trxn too
         $updateFinancialItemInfoValues['financialTrxn'] = $this->_getRelatedCancelFinancialTrxn($previousFinancialItemID);
         if ($previousLineItems[$updateFinancialItemInfoValues['entity_id']]['tax_amount']) {
-          $updateFinancialItemInfoValues['tax']['amount'] = -($previousLineItems[$updateFinancialItemInfoValues['entity_id']]['tax_amount']);
-          $updateFinancialItemInfoValues['tax']['description'] = $taxTerm;
-          if ($updateFinancialItemInfoValues['financial_type_id']) {
-            $updateFinancialItemInfoValues['tax']['financial_account_id'] = CRM_Contribute_BAO_Contribution::getFinancialAccountId($updateFinancialItemInfoValues['financial_type_id']);
-          }
+          $updateFinancialItemInfoValues['tax'] = $this->_getSalesTaxFinancialItem($previousLineItems[$updateFinancialItemInfoValues['entity_id']]);
+          $updateFinancialItemInfoValues['tax']['amount'] = -($updateFinancialItemInfoValues['tax']['amount']);
         }
+        // ensure that we are not creating duplicate entries
+        $params = array(
+          'amount' => $updateFinancialItemInfoValues['amount'],
+          'entity_table' => 'civicrm_line_item',
+          'entity_id' => $updateFinancialItemInfoValues['entity_id'],
+        );
         // INSERT negative financial_items for tax amount
-        $financialItemsArray[] = $updateFinancialItemInfoValues;
+        $financialItemsArray[$updateFinancialItemInfoValues['entity_id']] = $updateFinancialItemInfoValues;
       }
-      // if submitted and difference is 0 add a positive entry again
-      elseif (in_array($updateFinancialItemInfoValues['price_field_value_id'], $submittedPriceFieldValueIDs) && $updateFinancialItemInfoValues['differenceAmt'] == 0) {
-        $updateFinancialItemInfoValues['amount'] = $updateFinancialItemInfoValues['amount'];
-        // INSERT financial_items for tax amount
-        if ($updateFinancialItemInfoValues['entity_id'] == $lineItemsToUpdate[$updateFinancialItemInfoValues['price_field_value_id']]['id'] &&
-          isset($lineItemsToUpdate[$updateFinancialItemInfoValues['price_field_value_id']]['tax_amount'])
-        ) {
-          $updateFinancialItemInfoValues['tax']['amount'] = $lineItemsToUpdate[$updateFinancialItemInfoValues['price_field_value_id']]['tax_amount'];
-          $updateFinancialItemInfoValues['tax']['description'] = $taxTerm;
-          if ($lineItemsToUpdate[$updateFinancialItemInfoValues['price_field_value_id']]['financial_type_id']) {
-            $updateFinancialItemInfoValues['tax']['financial_account_id'] = CRM_Contribute_BAO_Contribution::getFinancialAccountId($lineItemsToUpdate[$updateFinancialItemInfoValues['price_field_value_id']]['financial_type_id']);
-          }
+      // if fee is changed to higher amount then update the existing financial item, this condition will hit ONLY when no payment has been made yet.
+      elseif (!empty($lineItemsToUpdate) &&
```

Comment:
this portion relates to tax - not assessed by me as yet

Reply:
Now only tax, this actually fixes the bug mentioned in "CRM-19273 : Change FeeSelection on pay-later" where earlier financial items are not updated to reflect the change.

Reply:
Just clarifying: no financial item should ever be changed, whether a payment has been made or not. If a price on a pay later is changed, that still requires the original financial item to remain there, and the change to be posted separately, in order to make the system auditable and compliant with accounting requirements.

Reply:
If we can get the problem resolved for everything except Sales Taxes, and there are still issues with Sales Taxes, I suggest we create a separate issue for that, merge this PR, and close the current issue.